apps/pipelines/nodes/nodes.py

```diff
@@ -94,9 +94,92 @@ def fn(input: Input):
+        session = self.experiment_session(state)
+        return session.get_participant_data()
+
+    @staticmethod
+    def to_json_schema(data: dict):
+        """Converts a dictionary to a JSON schema by first converting it to a Pydantic object and dumping it again.
+        The input should be in the format {"key": "description", "key2": [{"key": "description"}]}
+
+        Nested objects are not supported at the moment
+
+        Input example 1:
+        {"name": "the user's name", "surname": "the user's surname"}
+
+        Input example 2:
+        {"name": "the user's name", "pets": [{"name": "the pet's name": "type": "the type of animal"}]}
+
+        """
+
+        def _create_model_from_data(value_data, model_name: str):
+            pydantic_schema = {}
+            for key, value in value_data.items():
+                if isinstance(value, str):
+                    pydantic_schema[key] = (str | None, Field(description=value))
+                elif isinstance(value, list):
+                    model = _create_model_from_data(value[0], key.capitalize())
+                    pydantic_schema[key] = (list[model], Field(description=f"A list of {key}"))
+            return create_model(model_name, **pydantic_schema)
+
+        Model = _create_model_from_data(data, "CustomModel")
+        schema = Model.model_json_schema()
+        # The schema needs a description in order to comply with function calling APIs
+        schema["description"] = ""
+        return schema
+
+
+class UpdateParticipantMemory(PipelineNode):
+    """A simple component to merge the the input data into the participant's memory data. If key_name is specified
+    the input data will be merged with `key_name` as the key name.
+    """
+
+    __human_name__ = "Update participant memory"
+    key_name: str | None = None
+
+    def get_runnable(self, node: Node, state: PipelineState) -> RunnableLambda:
+        node = node
+
+        def fn(input: Input, config: RunnableConfig, state: PipelineState):
+            """Input should be a python dictionary"""
+            session = self.experiment_session(state)
+            extracted_data = {self.key_name: input} if self.key_name else input
+            try:
+                participant_data = ParticipantData.objects.for_experiment(session.experiment).get(
+                    participant=session.participant
+                )
+                participant_data.data = participant_data.data | extracted_data
+                participant_data.save()
+            except ParticipantData.DoesNotExist:
+                ParticipantData.objects.create(
+                    participant=session.participant,
+                    content_type__model="experiment",
+                    object_id=session.experiment.id,
+                    team=session.team,
+                    data=extracted_data,
+                )
+
+        return RunnableLambda(partial(fn, state=state), name=self.__class__.__name__)
```

Comment:
I don't think you need to use `partial` here. Isn't the `input` param to `fn` the state already?

Reply:
Hmm, on digging further, I see it isn't. But I think it should be. I don't think `apps.pipelines.nodes.base.PipelineNode.get_callable` should be manipulating the runnable input based on the type of runnable. It should just pass in the state and the runnable must do the right thing. Same with the output.